File: go-strats/pkg/strategy/registry.go
// Package strategy provides the strategy registry with all 100 strategy definitions.
//
// Mirrors the Python registry in src/strats_prob/registry.py and all
// strategy definitions in src/strats_prob/strategies/.
package strategy

import (
	"log/slog"
	"sort"
	"sync"

	"github.com/algomatic/strats100/go-strats/pkg/types"
)

var (
	mu            sync.RWMutex
	strategiesByID   = make(map[int]*types.StrategyDef)
	strategiesByName = make(map[string]*types.StrategyDef)
)

// Register adds a strategy to the registry.
func Register(s *types.StrategyDef) {
	mu.Lock()
	defer mu.Unlock()
	strategiesByID[s.ID] = s
	strategiesByName[s.Name] = s
}

// RegisterAll adds multiple strategies to the registry.
func RegisterAll(strategies []*types.StrategyDef) int {
	mu.Lock()
	defer mu.Unlock()
	count := 0
	for _, s := range strategies {
		strategiesByID[s.ID] = s
		strategiesByName[s.Name] = s
		count++
	}
	slog.Info("Registered strategies", "count", count, "total", len(strategiesByID))
	return count
}

// Get returns a strategy by ID, or nil if not found.
func Get(id int) *types.StrategyDef {
	mu.RLock()
	defer mu.RUnlock()
	return strategiesByID[id]
}

// GetByName returns a strategy by name, or nil if not found.
func GetByName(name string) *types.StrategyDef {
	mu.RLock()
	defer mu.RUnlock()
	return strategiesByName[name]
}

// GetAll returns all registered strategies sorted by ID.
func GetAll() []*types.StrategyDef {
	mu.RLock()
	defer mu.RUnlock()
	result := make([]*types.StrategyDef, 0, len(strategiesByID))
	for _, s := range strategiesByID {
		result = append(result, s)
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].ID < result[j].ID
	})
	return result
}

// GetByCategory returns all strategies of a given category sorted by ID.
func GetByCategory(category string) []*types.StrategyDef {
	mu.RLock()
	defer mu.RUnlock()
	var result []*types.StrategyDef
	for _, s := range strategiesByID {
		if s.Category == category {
			result = append(result, s)
		}
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].ID < result[j].ID
	})
	return result
}

// Clear removes all registered strategies (useful for testing).
func Clear() {
	mu.Lock()
	defer mu.Unlock()
	strategiesByID = make(map[int]*types.StrategyDef)
	strategiesByName = make(map[string]*types.StrategyDef)
}

// Count returns the number of registered strategies.
func Count() int {
	mu.RLock()
	defer mu.RUnlock()
	return len(strategiesByID)
}

File: src/strats_prob/registry.py
"""Strategy registry for probe system.

Maintains an in-memory registry of all strategy definitions and
optionally seeds them into the database.
"""

import logging
from typing import Optional

from src.strats_prob.strategy_def import StrategyDef

logger = logging.getLogger(__name__)

# Module-level registry
_strategies_by_id: dict[int, StrategyDef] = {}
_strategies_by_name: dict[str, StrategyDef] = {}


def register_strategies(strategies: list[StrategyDef]) -> int:
    """Register a list of strategy definitions.

    Args:
        strategies: List of StrategyDef instances to register.

    Returns:
        Number of strategies registered.
    """
    count = 0
    for s in strategies:
        # Detect ID collisions
        if s.id in _strategies_by_id:
            existing = _strategies_by_id[s.id]
            raise ValueError(
                f"Duplicate strategy ID {s.id}: "
                f"new='{s.name}' collides with existing='{existing.name}'"
            )
        # Detect name collisions
        if s.name in _strategies_by_name:
            existing = _strategies_by_name[s.name]
            raise ValueError(
                f"Duplicate strategy name '{s.name}': "
                f"new id={s.id} collides with existing id={existing.id}"
            )
        _strategies_by_id[s.id] = s
        _strategies_by_name[s.name] = s
        count += 1
    logger.info("Registered %d strategies (total: %d)", count, len(_strategies_by_id))
    return count


def get_strategy(strategy_id: int) -> Optional[StrategyDef]:
    """Get a strategy by ID."""
    return _strategies_by_id.get(strategy_id)


def get_strategy_by_name(name: str) -> Optional[StrategyDef]:
    """Get a strategy by name."""
    return _strategies_by_name.get(name)


def get_all_strategies() -> list[StrategyDef]:
    """Get all registered strategies sorted by ID."""
    return sorted(_strategies_by_id.values(), key=lambda s: s.id)


def get_strategies_by_category(category: str) -> list[StrategyDef]:
    """Get all registered strategies of a given category."""
    return sorted(
        [s for s in _strategies_by_id.values() if s.category == category],
        key=lambda s: s.id,
    )


def seed_strategies_to_db(session) -> int:
    """Seed all registered strategies into the probe_strategies DB table.

    Args:
        session: SQLAlchemy session.

    Returns:
        Number of rows affected.
    """
    from src.data.database.probe_repository import ProbeRepository

    repo = ProbeRepository(session)
    db_records = [s.to_db_dict() for s in get_all_strategies()]
    return repo.seed_strategies(db_records)


def clear_registry() -> None:
    """Clear all registered strategies (useful for testing)."""
    _strategies_by_id.clear()
    _strategies_by_name.clear()
